drivers/sdhc/sdhc_esp32.c

```diff
@@ -0,0 +1,1544 @@
+}
+
+/*
+ * Get host properties
+ */
+static int sdhc_esp32_get_host_props(const struct device *dev, struct sdhc_host_props *props)
+{
+	struct sdhc_esp32_data *data = dev->data;
+
+	memcpy(props, &data->props, sizeof(struct sdhc_host_props));
+	return 0;
+}
+
+/*
+ * Enable SDHC card interrupt
+ */
+static int sdhc_esp32_enable_interrupt(const struct device *dev, sdhc_interrupt_cb_t callback,
+				       int sources, void *user_data)
+{
+	return -ENOTSUP;
+}
+
+/*
+ * Disable SDHC card interrupt
+ */
+static int sdhc_esp32_disable_interrupt(const struct device *dev, int sources)
+{
+	return -ENOTSUP;
+}
+
+/**
+ * @brief SDMMC interrupt handler
+ *
+ * All communication in SD protocol is driven by the master, and the hardware
+ * handles things like stop commands automatically.
+ * So the interrupt handler doesn't need to do much, we just push interrupt
+ * status into a queue, clear interrupt flags, and let the task currently
+ * doing communication figure out what to do next.
+ * This also applies to SDIO interrupts which are generated by the slave.
+ *
+ * Card detect interrupts pose a small issue though, because if a card is
+ * plugged in and out a few times, while there is no task to process
+ * the events, event queue can become full and some card detect events
+ * may be dropped. We ignore this problem for now, since the there are no other
+ * interesting events which can get lost due to this.
+ */
+static void IRAM_ATTR sdio_esp32_isr(void *arg)
+{
+	sdmmc_dev_t *sdio_hw = s_host_ctx.hal.dev;
+	struct sdmmc_event event;
+	struct k_msgq *queue = (struct k_msgq *)arg;
+	uint32_t pending = sdmmc_ll_get_intr_status(sdio_hw) & 0xFFFF;
+
+	sdio_hw->rintsts.val = pending;
+	event.sdmmc_status = pending;
+
+	uint32_t dma_pending = sdio_hw->idsts.val;
+
+	sdio_hw->idsts.val = dma_pending;
+	event.dma_status = dma_pending & 0x1f;
+
+	if ((pending != 0) || (dma_pending != 0)) {
+		k_msgq_put(queue, &event, K_NO_WAIT);
+	}
+
+	uint32_t sdio_pending = (sdmmc_ll_get_intr_status(sdio_hw) &
+				 (SDMMC_INTMASK_IO_SLOT1 | SDMMC_INTMASK_IO_SLOT0));
+
+	if (sdio_pending) {
+		/*
+		 * Disable the interrupt (no need to clear here,
+		 * this is done in sdmmc_host_io_int_wait)
+		 */
+		sdmmc_ll_enable_interrupt(sdio_hw, sdio_pending, false);
+		k_sem_give(&s_host_ctx.io_intr_event);
+	}
+}
+
+/*
+ * Perform early system init for SDHC
+ */
+static int sdhc_esp32_init(const struct device *dev)
+{
+	const struct sdhc_esp32_config *cfg = dev->config;
+	struct sdhc_esp32_data *data = dev->data;
+	sdmmc_dev_t *sdio_hw = (sdmmc_dev_t *)cfg->sdio_hw;
+	int ret;
+
+	/* Pin configuration */
+
+	/* Set power GPIO high, so card starts powered */
+	if (cfg->pwr_gpio.port) {
+		ret = gpio_pin_configure_dt(&cfg->pwr_gpio, GPIO_OUTPUT_ACTIVE);
+
+		if (ret) {
+			return -EIO;
+		}
+	}
+
+#if CONFIG_SOC_SERIES_ESP32
+	/* SDMMC pins must be configured using IOMUX for these SOC models */
+	configure_pin_iomux(SDMMC_SLOT1_IOMUX_PIN_NUM_CLK);
```

Comment:
Not familiar with ESP32 pin control, but why can't these settings be applied via the `pinctrl` API?